tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+                                 EVP_PKEY_CTX * public_key_ctx,
+                                 EVP_PKEY_CTX * private_key_ctx,
+                                 unsigned char *id, size_t id_len,
+                                 unsigned char **session_key,
+                                 size_t *session_key_len)
+{
+  // Generate nonce B
+  unsigned char *nonce_b = NULL;
+  size_t nonce_b_len = 0;
+
+  int result = generate_nonce(32, &nonce_b, &nonce_b_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to generate a nonce.");
+    return 1;
+  }
+
+  // Conduct NSL to obtain nonce A
+  unsigned char *response = calloc(8192, sizeof(unsigned char));
+  size_t response_len = 8192;
+
+  struct sockaddr_storage peer_addr;
+  socklen_t peer_addr_len = sizeof(peer_addr);
+
+  ssize_t read_result = recvfrom(socket_fd,
+                                 response, response_len,
+                                 0,
+                                 (struct sockaddr *) &peer_addr,
+                                 &peer_addr_len);
+
+  if (-1 == read_result)
+  {
+    kmyth_log(LOG_ERR, "Failed to receive the nonce request.");
+    return 1;
+  }
+
+  char host[NI_MAXHOST] = { 0 };
+  char service[NI_MAXSERV] = { 0 };
+
+  int s = getnameinfo((struct sockaddr *) &peer_addr, peer_addr_len,
+                      host, NI_MAXHOST, service, NI_MAXSERV, NI_NUMERICSERV);
+
+  if (0 != s)
+  {
+    kmyth_log(LOG_ERR, "Failed to lookup host and service information.");
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Received %zd bytes from %s:%s", read_result, host,
+            service);
+
+  unsigned char *received_nonce_a = NULL;
+  size_t received_nonce_a_len = 0;
+
+  unsigned char *received_id = NULL;
+  size_t received_id_len = 0;
+
+  result = parse_nonce_request(private_key_ctx,
+                               response, read_result,
+                               &received_nonce_a, &received_nonce_a_len,
+                               &received_id, &received_id_len);
+
+  kmyth_log(LOG_INFO, "Received nonce A: %zd bytes", received_nonce_a_len);
+
+  free(response);
+  response = NULL;
+  response_len = 0;
+
+  kmyth_log(LOG_INFO, "Sending nonce B: %zd", nonce_b_len);
+
+  result = build_nonce_response(public_key_ctx,
+                                received_nonce_a, received_nonce_a_len,
+                                nonce_b, nonce_b_len,
+                                id, id_len, &response, &response_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to build the nonce response.");
+    return 1;
+  }
+
+  ssize_t send_result = sendto(socket_fd,
+                               response, response_len,
+                               0,
+                               (struct sockaddr *) &peer_addr, peer_addr_len);
+
+  if (response_len != send_result)
+  {
+    kmyth_log(LOG_ERR, "Failed to fully send the nonce response.");
+    return 1;
+  }
+
+  free(response);
+  response = calloc(8192, sizeof(unsigned char));
```

Comment:
We should check that the return value of calloc() is not NULL.

Reply:
Done